File: android/sdk/src/main/kotlin/com/atlassian/meridian/MeridianClient.kt
package com.atlassian.meridian

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

/**
 * Meridian API client for Kotlin/JVM and Android
 * Uses HttpURLConnection for universal JVM/Android compatibility
 * Coroutines for async operations with UI dispatcher where needed
 */
class MeridianClient(
  private val baseURL: String,
  private val sessionId: String,
  private val clientConfig: MeridianClientConfig = MeridianClientConfig(),
  val providerConfigMetrics: ProviderConfigMetrics = ProviderConfigMetrics(),
  val paymentEventLog: PaymentEventLog = PaymentEventLog(),
  private val clock: () -> Long = System::currentTimeMillis,
) {
  private val mapper = jsonMapper()
  private val baseUrlNormalized = baseURL.removeSuffix("/")
  private val providerConfigCache = ProviderConfigCache(
    ttlMillis = clientConfig.providerConfigTtlMillis,
    clock = clock,
  )

  init {
    require(sessionId.isNotEmpty()) { "Session ID is required" }
    require(
      baseUrlNormalized.startsWith("http://") || baseUrlNormalized.startsWith("https://")
    ) { "Invalid URL: must start with http:// or https://" }
  }

  // MARK: - Internal Request Method

  private suspend fun <T> request(
    method: String,
    path: String,
    body: Any? = null,
    additionalHeaders: Map<String, String> = emptyMap(),
    responseType: Class<T>,
  ): T = withContext(Dispatchers.IO) {
    val fullUrl = "$baseUrlNormalized$path"
    val url = URL(fullUrl)

    val connection = url.openConnection() as HttpURLConnection
    try {
      connection.connectTimeout = 15000
      connection.readTimeout = 15000
      connection.requestMethod = method
      connection.setRequestProperty("Content-Type", "application/json")
      connection.setRequestProperty("X-Rehearsal-Session", sessionId)

      // Add additional headers (e.g., Idempotency-Key)
      additionalHeaders.forEach { (key, value) ->
        connection.setRequestProperty(key, value)
      }

      // Write body if present
      if (body != null) {
        val bodyJson = mapper.writeValueAsString(body)
        connection.doOutput = true
        connection.outputStream.use { out ->
          out.write(bodyJson.toByteArray(StandardCharsets.UTF_8))
          out.flush()
        }
      }

      // Read response
      val statusCode = connection.responseCode
      val responseStream = if (statusCode >= 400) {
        connection.errorStream
      } else {
        connection.inputStream
      }

      val responseBody = responseStream?.bufferedReader()?.use { it.readText() } ?: ""

      // Check HTTP status and handle errors
      when {
        statusCode >= 200 && statusCode < 300 -> {
          // Success
          try {
            mapper.readValue(responseBody, responseType)
          } catch (e: Exception) {
            throw MeridianError.DecodingError("Failed to parse response: ${e.message}", e)
          }
        }

        statusCode == 202 || statusCode == 400 || statusCode == 409 || statusCode == 422 || statusCode == 503 -> {
          // These are expected error statuses, parse the response
          try {
            mapper.readValue(responseBody, responseType)
          } catch (e: Exception) {
            throw MeridianError.HttpError(
              statusCode,
              responseBody.ifEmpty { "Unknown error" }
            )
          }
        }

        else -> {
          throw MeridianError.HttpError(
            statusCode,
            responseBody.ifEmpty { "Unknown error" }
          )
        }
      }
    } finally {
      connection.disconnect()
    }
  }

  // MARK: - Public API Methods

  /**
   * GET /health - Check service health
   */
  suspend fun getHealth(): HealthResponse =
    request("GET", "/health", responseType = HealthResponse::class.java)

  /**
   * GET /catalog - Fetch recipients and providers
   */
  suspend fun getCatalog(): CatalogResponse =
    request("GET", "/catalog", responseType = CatalogResponse::class.java)

  /**
   * Payment methods for the picker.
   * With [MeridianClientConfig.configDrivenProviders] off, returns the hardcoded
   * Adyen card and Worldpay bank list and does not fetch configuration.
   * With the flag on, loads options from GET /catalog, serves a fresh in-memory
   * cache, and on timeout or failure returns the last-known-good list or that
   * same two-provider baseline. The selected provider is never swapped for a
   * different one because a fetch timed out.
   */
  suspend fun paymentMethodOptions(): List<PaymentMethodOption> {
    if (!clientConfig.configDrivenProviders) {
      return ProviderCatalog.baseline()
    }
    providerConfigCache.fresh()?.let { return it }
    val fetched = try {
      fetchProviderOptions()
    } catch (e: CancellationException) {
      throw e
    } catch (e: Exception) {
      CatalogFetch(null, null, e.message ?: "catalog fetch failed")
    }
    val options = fetched.options
    if (options != null) {
      providerConfigCache.store(options)
      providerConfigMetrics.recordFetchSuccess()
      return options
    }
    val failure = ProviderConfigFailure(
      correlationId = fetched.correlationId,
      sessionId = sessionId,
      reason = fetched.reason ?: "catalog fetch failed",
    )
    providerConfigMetrics.recordFetchFailure(failure)
    val known = providerConfigCache.lastKnown()
    if (known != null) {
      providerConfigMetrics.recordFallbackToCache(failure)
      return known
    }
    return ProviderCatalog.baseline()
  }

  private data class CatalogFetch(
    val options: List<PaymentMethodOption>?,
    val correlationId: String?,
    val reason: String?,
  )

  private suspend fun fetchProviderOptions(): CatalogFetch = withContext(Dispatchers.IO) {
    if (!ProviderCatalog.configTransportAllowed(baseUrlNormalized)) {
      return@withContext CatalogFetch(
        options = null,
        correlationId = null,
        reason = "Provider configuration requires HTTPS",
      )
    }
    val connection = try {
      URL("$baseUrlNormalized/catalog").openConnection() as HttpURLConnection
    } catch (e: CancellationException) {
      throw e
    } catch (e: Exception) {
      return@withContext CatalogFetch(null, null, e.message ?: "catalog fetch failed")
    }
    try {
      connection.instanceFollowRedirects = false
      connection.connectTimeout = clientConfig.providerConfigTimeoutMillis
      connection.readTimeout = clientConfig.providerConfigTimeoutMillis
      connection.requestMethod = "GET"
      connection.setRequestProperty("Content-Type", "application/json")
      connection.setRequestProperty("X-Rehearsal-Session", sessionId)
      val statusCode = connection.responseCode
      val responseStream = if (statusCode >= 400) connection.errorStream else connection.inputStream
      val responseBody = responseStream?.bufferedReader()?.use { it.readText() } ?: ""
      val headers = connection.headerFields
        ?.mapNotNull { (key, value) -> if (key == null) null else key to value.toList() }
        ?.toMap()
        ?: emptyMap()
      val correlationId = ProviderCatalog.correlationId(headers, responseBody, mapper)
      if (statusCode !in 200..299) {
        return@withContext CatalogFetch(null, correlationId, "HTTP $statusCode")
      }
      val catalog = try {
        mapper.readValue(responseBody, CatalogResponse::class.java)
      } catch (e: Exception) {
        return@withContext CatalogFetch(null, correlationId, "Malformed catalog")
      }
      val options = ProviderCatalog.optionsFromCatalog(catalog)
      if (options.isEmpty()) {
        return@withContext CatalogFetch(null, correlationId, "Catalog did not include a live provider")
      }
      CatalogFetch(options, correlationId, null)
    } catch (e: CancellationException) {
      throw e
    } catch (e: Exception) {
      CatalogFetch(null, null, e.message ?: "catalog fetch failed")
    } finally {
      connection.disconnect()
    }
  }

  /**
   * GET /state - Fetch current bank state
   */
  suspend fun getState(): BankState =
    request("GET", "/state", responseType = BankState::class.java)

  /**
   * POST /payments - Submit a payment
   * @param recipientId Recipient ID
   * @param amountMinor Amount in GBP pence (integer)
   * @param method Payment method (card or bank)
   * @param note Optional note (max 200 chars)
   * @param scenario Simulation scenario
   * @param idempotencyKey Unique key for idempotency
   */
  suspend fun submitPayment(
    recipientId: String,
    amountMinor: Int,
    method: PaymentMethod,
    note: String = "",
    scenario: Scenario = Scenario.success,
    idempotencyKey: String,
  ): PaymentResponse {
    val selection = ProviderCatalog.baseline().first { ProviderCatalog.wireMethod(it.id) == method }
    return postPayment(recipientId, amountMinor, selection, method, note, scenario, idempotencyKey)
  }

  /**
   * Submit a payment using a catalog method id (`adyen_card` or `worldpay_bank`).
   * The Idempotency-Key header and the wire `method` (`card` or `bank`) are unchanged.
   * An unknown id is rejected locally and is not sent to another provider.
   */
  suspend fun submitPayment(
    recipientId: String,
    amountMinor: Int,
    methodId: String,
    note: String = "",
    scenario: Scenario = Scenario.success,
    idempotencyKey: String,
  ): PaymentResponse {
    val resolved = ProviderCatalog.resolve(methodId)
      ?: throw MeridianError.ValidationError("Unknown payment method")
    val selection = providerConfigCache.lastKnown()
      ?.firstOrNull { it.id == resolved.option.id }
      ?: resolved.option
    return postPayment(
      recipientId,
      amountMinor,
      selection,
      resolved.method,
      note,
      scenario,
      idempotencyKey,
    )
  }

  private suspend fun postPayment(
    recipientId: String,
    amountMinor: Int,
    selection: PaymentMethodOption,
    method: PaymentMethod,
    note: String,
    scenario: Scenario,
    idempotencyKey: String,
  ): PaymentResponse {
    val payload = PaymentRequest(
      recipientId = recipientId,
      amountMinor = amountMinor,
      method = method.name,
      note = note,
      scenario = scenario.name,
    )
    try {
      val response = request(
        "POST",
        "/payments",
        payload,
        mapOf("Idempotency-Key" to idempotencyKey),
        PaymentResponse::class.java,
      )
      paymentEventLog.record(
        PaymentTransactionEvent(
          methodId = selection.id,
          providerName = selection.providerName,
          wireMethod = method.name,
          idempotencyKey = idempotencyKey,
          transactionId = response.transaction?.id ?: response.paymentId,
          code = response.code,
        )
      )
      return response
    } catch (e: Exception) {
      paymentEventLog.record(
        PaymentTransactionEvent(
          methodId = selection.id,
          providerName = selection.providerName,
          wireMethod = method.name,
          idempotencyKey = idempotencyKey,
          transactionId = null,
          code = "UNKNOWN",
        )
      )
      throw e
    }
  }

  /**
   * PATCH /budgets - Update budget for a category
   * @param category Budget category
   * @param limitMinor Limit in GBP pence
   */
  suspend fun updateBudget(
    category: String,
    limitMinor: Int,
  ): BudgetResponse {
    val payload = BudgetRequest(category = category, limitMinor = limitMinor)
    return request(
      "PATCH",
      "/budgets",
      payload,
      responseType = BudgetResponse::class.java,
    )
  }

  /**
   * POST /reset - Reset session state
   */
  suspend fun reset(): ResetResponse =
    request("POST", "/reset", responseType = ResetResponse::class.java)

  /**
   * GET /events - Fetch audit events
   */
  suspend fun getEvents(): EventsResponse =
    request("GET", "/events", responseType = EventsResponse::class.java)

  companion object {
    /**
     * Unknown JSON fields are ignored so a catalog extension from meridian-api
     * does not fail the payment flow before the contract is reconciled.
     */
    internal fun jsonMapper(): ObjectMapper =
      ObjectMapper().registerKotlinModule().apply {
        configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
      }
  }
}

File: ios/MeridianSDK/Sources/MeridianSDK/MeridianClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public actor MeridianClient {
  private let baseURL: URL
  private let sessionId: String
  private let transport: any MeridianHTTPTransport
  private let features: MeridianFeatures
  private let providerCacheTTL: TimeInterval
  private let providerFetchTimeout: TimeInterval
  private let now: @Sendable () -> Date
  private let metricsHandler: (@Sendable (ProviderConfigMetricEvent) -> Void)?
  private var providerCache: (options: [PaymentMethodOption], fetchedAt: Date)?
  private var metricEvents: [ProviderConfigMetricEvent] = []
  private var paymentLogs: [PaymentTransactionLog] = []

  /// Initialize Meridian API client
  /// - Parameters:
  ///   - baseURL: API base URL (e.g., "http://localhost:8080/api/v1" or an HTTPS host)
  ///   - sessionId: Rehearsal session ID (3-64 URL-safe ASCII)
  ///   - urlSession: Optional URLSession for testing
  ///   - features: Feature flags. Config-driven providers default off.
  ///   - providerCacheTTL: In-memory lifetime for a successful provider list.
  ///   - providerFetchTimeout: Catalog fetch timeout before last-known-good fallback.
  ///   - metricsHandler: Receives provider-config metric events as they are emitted.
  ///   - now: Clock used for the provider cache TTL.
  public init(
    baseURL: String,
    sessionId: String,
    urlSession: URLSession = .shared,
    features: MeridianFeatures = MeridianFeatures(),
    providerCacheTTL: TimeInterval = 30,
    providerFetchTimeout: TimeInterval = 3,
    metricsHandler: (@Sendable (ProviderConfigMetricEvent) -> Void)? = nil,
    now: @escaping @Sendable () -> Date = { Date() }
  ) throws {
    try self.init(
      validatedBaseURL: Self.validatedBaseURL(baseURL, sessionId: sessionId),
      sessionId: sessionId,
      transport: URLSessionTransport(session: urlSession),
      features: features,
      providerCacheTTL: providerCacheTTL,
      providerFetchTimeout: providerFetchTimeout,
      metricsHandler: metricsHandler,
      now: now
    )
  }

  init(
    baseURL: String,
    sessionId: String,
    transport: any MeridianHTTPTransport,
    features: MeridianFeatures = MeridianFeatures(),
    providerCacheTTL: TimeInterval = 30,
    providerFetchTimeout: TimeInterval = 3,
    metricsHandler: (@Sendable (ProviderConfigMetricEvent) -> Void)? = nil,
    now: @escaping @Sendable () -> Date = { Date() }
  ) throws {
    try self.init(
      validatedBaseURL: Self.validatedBaseURL(baseURL, sessionId: sessionId),
      sessionId: sessionId,
      transport: transport,
      features: features,
      providerCacheTTL: providerCacheTTL,
      providerFetchTimeout: providerFetchTimeout,
      metricsHandler: metricsHandler,
      now: now
    )
  }

  private init(
    validatedBaseURL: URL,
    sessionId: String,
    transport: any MeridianHTTPTransport,
    features: MeridianFeatures,
    providerCacheTTL: TimeInterval,
    providerFetchTimeout: TimeInterval,
    metricsHandler: (@Sendable (ProviderConfigMetricEvent) -> Void)?,
    now: @escaping @Sendable () -> Date
  ) {
    self.baseURL = validatedBaseURL
    self.sessionId = sessionId
    self.transport = transport
    self.features = features
    self.providerCacheTTL = providerCacheTTL
    self.providerFetchTimeout = providerFetchTimeout
    self.metricsHandler = metricsHandler
    self.now = now
  }

  private static func validatedBaseURL(_ baseURL: String, sessionId: String) throws -> URL {
    guard !sessionId.isEmpty else {
      throw MeridianError.missingSession
    }

    // Normalize baseURL: remove trailing slash
    var urlStr = baseURL
    while urlStr.hasSuffix("/") {
      urlStr.removeLast()
    }

    guard !urlStr.isEmpty && urlStr.starts(with: "http://") || urlStr.starts(with: "https://") else {
      throw MeridianError.invalidURL
    }

    guard let url = URL(string: urlStr) else {
      throw MeridianError.invalidURL
    }
    return url
  }

  // MARK: - Internal Request Method

  private func perform(
    method: String,
    path: String,
    body: Encodable? = nil,
    additionalHeaders: [String: String] = [:],
    timeout: TimeInterval? = nil
  ) async throws -> (Data, HTTPURLResponse) {
    let url = baseURL.appendingPathComponent(path)

    var request = URLRequest(url: url)
    request.httpMethod = method
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue(sessionId, forHTTPHeaderField: "X-Rehearsal-Session")
    if let timeout {
      request.timeoutInterval = timeout
      request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
    }

    // Add additional headers (e.g., Idempotency-Key)
    for (key, value) in additionalHeaders {
      request.setValue(value, forHTTPHeaderField: key)
    }

    // Encode body if present
    if let body {
      let encoder = JSONEncoder()
      request.httpBody = try encoder.encode(body)
    }

    let (data, response) = try await transport.data(for: request)

    guard let httpResponse = response as? HTTPURLResponse else {
      throw MeridianError.networkError("Invalid response type")
    }

    return (data, httpResponse)
  }

  private func request<T: Decodable>(
    method: String,
    path: String,
    body: Encodable? = nil,
    additionalHeaders: [String: String] = [:]
  ) async throws -> T {
    let (data, httpResponse) = try await perform(
      method: method,
      path: path,
      body: body,
      additionalHeaders: additionalHeaders
    )

    // Check HTTP status
    guard httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 || httpResponse.statusCode == 202 || httpResponse.statusCode == 400 || httpResponse.statusCode == 409 || httpResponse.statusCode == 422 || httpResponse.statusCode == 503
    else {
      let errorMsg = String(data: data, encoding: .utf8) ?? "Unknown error"
      throw MeridianError.httpError(
        statusCode: httpResponse.statusCode,
        message: errorMsg
      )
    }

    let decoder = JSONDecoder()
    do {
      return try decoder.decode(T.self, from: data)
    } catch {
      throw MeridianError.decodingError(error.localizedDescription)
    }
  }

  // MARK: - Public API Methods

  /// GET /health - Check service health
  public func getHealth() async throws -> HealthResponse {
    return try await request(method: "GET", path: "/health")
  }

  /// GET /catalog - Fetch recipients and providers
  public func getCatalog() async throws -> CatalogResponse {
    return try await request(method: "GET", path: "/catalog")
  }

  /// Provider methods for the picker.
  ///
  /// With `configDrivenProviders` off, this is the hardcoded Adyen card and Worldpay bank list
  /// and does not touch the network. With the flag on, the list is read from `GET /catalog`
  /// over the configured base URL (HTTPS for a shared host), cached for a short TTL, and
  /// replaced by the last-known-good list — or the same two-provider baseline — if the fetch
  /// fails, times out, or is malformed. This path does not throw.
  public func paymentMethodOptions() async -> [PaymentMethodOption] {
    if !features.configDrivenProviders {
      return ProviderConfiguration.baseline
    }
    if let providerCache, now().timeIntervalSince(providerCache.fetchedAt) < providerCacheTTL {
      return providerCache.options
    }

    switch await fetchProviderCatalog() {
    case let .ready(options, correlationId):
      providerCache = (options, now())
      emit(.fetchSuccess, correlationId: correlationId, reason: nil)
      return options
    case let .failed(reason, correlationId):
      emit(.fetchFailure, correlationId: correlationId, reason: reason)
      if let providerCache {
        emit(.fallbackToCache, correlationId: correlationId, reason: reason)
        return providerCache.options
      }
      emit(.fallbackToBaseline, correlationId: correlationId, reason: reason)
      return ProviderConfiguration.baseline
    }
  }

  public func providerConfigMetrics() -> ProviderConfigMetrics {
    ProviderConfigMetrics(
      fetchSuccess: metricEvents.filter { $0.kind == .fetchSuccess }.count,
      fetchFailure: metricEvents.filter { $0.kind == .fetchFailure }.count,
      fallbackToCache: metricEvents.filter { $0.kind == .fallbackToCache }.count,
      fallbackToBaseline: metricEvents.filter { $0.kind == .fallbackToBaseline }.count,
      events: metricEvents
    )
  }

  public func paymentTransactionLogs() -> [PaymentTransactionLog] {
    paymentLogs
  }

  /// GET /state - Fetch current bank state
  public func getState() async throws -> BankState {
    return try await request(method: "GET", path: "/state")
  }

  /// POST /payments - Submit a payment
  /// - Parameters:
  ///   - recipientId: Recipient ID
  ///   - amountMinor: Amount in GBP pence (integer)
  ///   - method: Payment method (card or bank)
  ///   - note: Optional note (max 200 chars)
  ///   - scenario: Simulation scenario
  ///   - idempotencyKey: Unique key for idempotency. Retained as supplied; never rewritten.
  public func submitPayment(
    recipientId: String,
    amountMinor: Int,
    method: PaymentMethod,
    note: String = "",
    scenario: Scenario = .success,
    idempotencyKey: String
  ) async throws -> PaymentResponse {
    guard let option = ProviderConfiguration.baseline.first(where: { $0.method == method }) else {
      throw MeridianError.validationError("Unknown payment method")
    }
    return try await performSubmit(
      recipientId: recipientId,
      amountMinor: amountMinor,
      option: option,
      note: note,
      scenario: scenario,
      idempotencyKey: idempotencyKey
    )
  }

  /// POST /payments using a provider-method id from `paymentMethodOptions()`.
  ///
  /// The idempotency key is forwarded unchanged. The JSON `method` field stays `card` or `bank`;
  /// the option id is not substituted into the body or the key. A timeout does not select a different provider.
  public func submitPayment(
    recipientId: String,
    amountMinor: Int,
    methodId: String,
    note: String = "",
    scenario: Scenario = .success,
    idempotencyKey: String
  ) async throws -> PaymentResponse {
    let options = await paymentMethodOptions()
    guard let option = options.first(where: { $0.id == methodId }) else {
      throw MeridianError.validationError("Unknown payment method")
    }
    return try await performSubmit(
      recipientId: recipientId,
      amountMinor: amountMinor,
      option: option,
      note: note,
      scenario: scenario,
      idempotencyKey: idempotencyKey
    )
  }

  /// PATCH /budgets - Update budget for a category
  /// - Parameters:
  ///   - category: Budget category
  ///   - limitMinor: Limit in GBP pence
  public func updateBudget(
    category: Category,
    limitMinor: Int
  ) async throws -> BudgetResponse {
    let payload = BudgetRequest(category: category, limitMinor: limitMinor)
    return try await request(
      method: "PATCH",
      path: "/budgets",
      body: payload
    )
  }

  /// POST /reset - Reset session state
  public func reset() async throws -> ResetResponse {
    return try await request(method: "POST", path: "/reset")
  }

  /// GET /events - Fetch audit events
  public func getEvents() async throws -> EventsResponse {
    return try await request(method: "GET", path: "/events")
  }

  // MARK: - Provider catalog

  private enum CatalogFetchResult {
    case ready([PaymentMethodOption], correlationId: String?)
    case failed(reason: String, correlationId: String?)
  }

  private func fetchProviderCatalog() async -> CatalogFetchResult {
    do {
      let (data, http) = try await perform(
        method: "GET",
        path: "/catalog",
        timeout: providerFetchTimeout
      )
      let correlationId = Self.correlationIdentifier(from: http, body: data)
      guard (200..<300).contains(http.statusCode) else {
        return .failed(reason: Self.httpFailureReason(http.statusCode, body: data), correlationId: correlationId)
      }
      let catalog: CatalogResponse
      do {
        catalog = try JSONDecoder().decode(CatalogResponse.self, from: data)
      } catch {
        return .failed(reason: "malformed-catalog", correlationId: correlationId)
      }
      guard let options = ProviderConfiguration.options(from: catalog) else {
        return .failed(reason: "incomplete-catalog", correlationId: correlationId)
      }
      return .ready(options, correlationId: correlationId)
    } catch {
      return .failed(reason: Self.isTimeout(error) ? "timeout" : "network", correlationId: nil)
    }
  }

  static func correlationIdentifier(from response: HTTPURLResponse, body: Data) -> String? {
    let headerNames = [
      "x-correlation-id",
      "x-request-id",
      "x-meridian-correlation-id",
      "x-meridian-request-id",
    ]
    for (rawKey, rawValue) in response.allHeaderFields {
      let name = String(describing: rawKey).trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
      guard headerNames.contains(name) else { continue }
      let text = String(describing: rawValue).trimmingCharacters(in: .whitespacesAndNewlines)
      if !text.isEmpty { return text }
    }
    guard let json = try? JSONSerialization.jsonObject(with: body) as? [String: Any] else { return nil }
    for key in ["correlationId", "requestId", "correlation_id", "request_id"] {
      if let value = json[key] as? String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmed.isEmpty { return trimmed }
      }
    }
    return nil
  }

  private static func httpFailureReason(_ status: Int, body: Data) -> String {
    if let json = try? JSONSerialization.jsonObject(with: body) as? [String: Any],
       let code = json["code"] as? String {
      let trimmed = code.trimmingCharacters(in: .whitespacesAndNewlines)
      if !trimmed.isEmpty {
        return "http-\(status) code=\(trimmed)"
      }
    }
    return "http-\(status)"
  }

  private static func isTimeout(_ error: Error) -> Bool {
    if let urlError = error as? URLError, urlError.code == .timedOut {
      return true
    }
    let nsError = error as NSError
    return nsError.domain == NSURLErrorDomain && nsError.code == NSURLErrorTimedOut
  }

  private func emit(_ kind: ProviderConfigMetricEvent.Kind, correlationId: String?, reason: String?) {
    let event = ProviderConfigMetricEvent(
      kind: kind,
      correlationId: correlationId,
      sessionId: sessionId,
      reason: reason
    )
    metricEvents.append(event)
    metricsHandler?(event)
  }

  private func performSubmit(
    recipientId: String,
    amountMinor: Int,
    option: PaymentMethodOption,
    note: String,
    scenario: Scenario,
    idempotencyKey: String
  ) async throws -> PaymentResponse {
    let payload = PaymentRequest(
      recipientId: recipientId,
      amountMinor: amountMinor,
      method: option.method,
      note: note,
      scenario: scenario
    )
    do {
      let response: PaymentResponse = try await request(
        method: "POST",
        path: "/payments",
        body: payload,
        additionalHeaders: ["Idempotency-Key": idempotencyKey]
      )
      paymentLogs.append(transactionLog(option: option, idempotencyKey: idempotencyKey, response: response))
      return response
    } catch {
      paymentLogs.append(transactionLog(option: option, idempotencyKey: idempotencyKey, response: nil))
      throw error
    }
  }

  private func transactionLog(
    option: PaymentMethodOption,
    idempotencyKey: String,
    response: PaymentResponse?
  ) -> PaymentTransactionLog {
    PaymentTransactionLog(
      action: "payment.submitted",
      idempotencyKey: idempotencyKey,
      methodId: option.id,
      providerName: option.providerName,
      method: option.method,
      transactionId: response?.transaction?.id,
      paymentId: response?.paymentId ?? response?.transaction?.id,
      reference: response?.transaction?.reference,
      sessionId: sessionId
    )
  }
}
